storage2/src/state.rs

```diff
@@ -38,23 +41,74 @@ impl State {
-        // If the key is available in the sidecar cache, return it.
-        if let Some(v) = self.sidecar_changes.get(key) {
+    async fn get_nonconsensus(&self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        // If the key is available in the nonconsensus cache, return it.
+        if let Some(v) = self.nonconsensus_changes.get(key) {
             return Ok(v.clone());
         }
 
         // Otherwise, if the key is available in the snapshot, return it.
-        self.snapshot.get_sidecar(key)
+        self.snapshot.get_nonconsensus(key).await
+    }
+
+    async fn prefix_raw(
+        &self,
+        prefix: &str,
+    ) -> Result<Pin<Box<dyn Stream<Item = (String, Box<[u8]>)> + Send + '_>>> {
+        // Interleave the unwritten_changes cache with the snapshot.
+        let (tx, rx) = mpsc::channel(100);
+
+        let mut snapshotted_stream = self.snapshot.prefix_raw(prefix).await?;
+        let mut snapshotted_match = snapshotted_stream.next().await;
+        for (key, value) in self.unwritten_changes.iter() {
+            // Iterate the unwritten_changes cache (sorted by key) until we reach the keys
+            // that match the prefix.
+            //
+            // If value is `None`, then the key has been deleted, and we should skip it.
+            if !key.starts_with(prefix) || value.is_none() {
+                continue;
+            }
+
+            let value = value.clone().unwrap();
+
+            // This key matches the prefix.
+            // While the snapshot prefix stream returns keys that lexicographically precede this key,
+            // return those.
+            while let Some((snapshotted_key, snapshotted_value)) = snapshotted_match {
+                if &snapshotted_key < key {
+                    // The snapshot key is less than the unwritten_changes key, so return
+                    // the snapshot key.
+                    tx.send((snapshotted_key, snapshotted_value)).await?;
+                    // And then advance the snapshot stream to the next match.
+                    snapshotted_match = snapshotted_stream.next().await;
+                } else {
+                    // Keep this match around for another iteration.
+                    snapshotted_match = Some((snapshotted_key, snapshotted_value));
+                }
+            }
+
+            // All snapshot matches preceding this unwritten_changes key have been sent to the channel,
+            // so send this key.
+            tx.send((key.to_string(), value.into_boxed_slice())).await?;
+        }
+
+        // Send any remaining data from the snapshot stream.
+        while let Some((snapshotted_key, snapshotted_value)) = snapshotted_match {
+            tx.send((snapshotted_key, snapshotted_value)).await?;
+            // Advance the snapshot stream to the next match.
+            snapshotted_match = snapshotted_stream.next().await;
+        }
+
+        Ok(Box::pin(tokio_stream::wrappers::ReceiverStream::new(rx)))
```

Comment:
I think this loop could be rewritten more simply using `BTreeMap::range` as noted above and [`IterTools::merge_by`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.merge_by) to merge the two iterators in order. Make sure to cut off the `BTreeMap::range` when it exceeds the prefix, perhaps using `.take_while()`?

Reply:
Turns out we can't use `IterTools::merge_by` because we don't have two iterators, and it's not clear that re-implementing that logic for `Stream` is worthwhile